src/compiler/checker.ts

```diff
@@ -8016,6 +8154,65 @@ module ts {
             }
         }
 
+        function getAwaitedType(type: Type, fallbackType?: Type, seen: boolean[] = []): Type {
+            var seen: boolean[] = [];
+            if (!type || !globalIPromiseType || !(type.flags & TypeFlags.ObjectType)) {
+                return fallbackType;
+            }
+            seen[type.id] = true;
+            if (checkTypeRelatedTo(type, globalIPromiseType, assignableRelation, undefined)) {
+                var thenProp = getPropertyOfType(type, "then");
+                var thenType = getTypeOfSymbol(thenProp);
+                var thenSignatures = getSignaturesOfType(thenType, SignatureKind.Call);
+                var onFulfilledParameterType = getUnionType(map(thenSignatures, signature => getTypeAtPosition(getErasedSignature(signature), 0)));
+                var onFulfilledParameterSignatures = getSignaturesOfType(onFulfilledParameterType, SignatureKind.Call);
+                var awaitedType = getUnionType(map(onFulfilledParameterSignatures, signature => getTypeAtPosition(getErasedSignature(signature), 0)));
+                if (seen[awaitedType.id]) return fallbackType;
+                return getAwaitedType(awaitedType, /*fallbackType*/ awaitedType, seen);
+            } else {
+                if (isTypeAssignableTo(type, thenableType)) {
+                    error(null, ts.Diagnostics.Type_for_await_does_not_have_a_valid_callable_then_member);
+                }
+                return fallbackType;
+            }
+        }
+
+        function checkAwaitableReturnType(node: SignatureDeclaration, returnType: Type): Type {
+            // an async function has a valid return type if the type has a construct signature that takes
+            // in an `initializer` function that in turn supplies a `resolve` function as one of its arguments
+            // and results in an object with a callable `then` signature.
+            if (globalIPromiseConstructorType) {

```

Comment:
There is an error reported if it’s not found, during initializeTypeChecker. I had this test in before some other refactoring and may no longer need it.

From: Jason Freeman [mailto:notifications@github.com]
Sent: Friday, February 13, 2015 5:37 PM
To: Microsoft/TypeScript
Cc: Ron Buckton
Subject: Re: [TypeScript] Async Functions (#1781)


In src/compiler/checker.ts<https://github.com/Microsoft/TypeScript/pull/1781#discussion_r24710362>:

> +                var awaitedType = getUnionType(map(onFulfilledParameterSignatures, signature => getTypeAtPosition(getErasedSignature(signature), 0)));

> +                if (seen[awaitedType.id]) return fallbackType;

> +                return getAwaitedType(awaitedType, /*fallbackType*/ awaitedType, seen);

> +            } else {

> +                if (isTypeAssignableTo(type, thenableType)) {

> +                    error(null, ts.Diagnostics.Type_for_await_does_not_have_a_valid_callable_then_member);

> +                }

> +                return fallbackType;

> +            }

> +        }

> +

> +        function checkAwaitableReturnType(node: SignatureDeclaration, returnType: Type): Type {

> +            // an async function has a valid return type if the type has a construct signature that takes

> +            // in an `initializer` function that in turn supplies a `resolve` function as one of its arguments

> +            // and results in an object with a callable `then` signature.

> +            if (globalIPromiseConstructorType) {

First, I would invert this check, so you return if it is not found. Second, I would prefer to give an error that the type was not found, when you first try to look for the global type, rather than penalize all async functions.

—
Reply to this email directly or view it on GitHub<https://github.com/Microsoft/TypeScript/pull/1781/files#r24710362>.